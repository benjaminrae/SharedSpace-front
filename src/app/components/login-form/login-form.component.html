<form [formGroup]="loginForm" (ngSubmit)="onSubmit()">
  <h2>Log in to SharedSpace</h2>

  <div class="form-group">
    <label for="username">Username</label>
    <input id="username" type="text" formControlName="username">
  </div>

  <div class="form-group">
    <label for="password">Password</label>
    <input id="password" type="password" formControlName="password">
  </div>

  <app-button text="Login" [big]="true" [disabled]="loginForm.status === 'INVALID'"></app-button>

  <span>Don't have an account? <a href="">Register here</a> </span>

</form>
