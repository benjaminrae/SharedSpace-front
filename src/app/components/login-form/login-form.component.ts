import { Component } from "@angular/core";
import { Validators, FormBuilder } from "@angular/forms";
import { TokenService } from "src/app/services/token/token.service";
import { UiService } from "src/app/services/ui/ui.service";
import { UserService } from "src/app/services/user/user.service";
import { UserCredentials } from "src/app/store/user-feature/types";

@Component({
  selector: "app-login-form",
  templateUrl: "./login-form.component.html",
})
export class LoginFormComponent {
  loginForm = this.fb.group({
    username: ["", [Validators.required, Validators.minLength(5)]],
    password: ["", [Validators.required, Validators.minLength(8)]],
  });

  constructor(
    private readonly fb: FormBuilder,
    private readonly userService: UserService,
    private readonly uiService: UiService,
    private readonly tokenService: TokenService
  ) {}

  onSubmit() {
    this.uiService.showLoading();

    const data = this.userService.getToken(
      this.loginForm.value as UserCredentials
    );

    data.subscribe((data) => {
      const { token } = data;
      const { id, username } = this.tokenService.decodeToken(token);
      this.userService.loginUser({ username, id, token });
      this.uiService.hideLoading();
      this.resetForm();
      this.uiService.showSuccessModal("You have logged in correctly");
    });
  }

  private resetForm() {
    this.loginForm.reset();
  }
}
