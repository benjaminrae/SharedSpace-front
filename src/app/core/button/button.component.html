<button [routerLink]="routerLink" [type]="type" [className]="className" (click)="onClickButton($event)"
  [disabled]="disabled" #button>{{text}}</button>
