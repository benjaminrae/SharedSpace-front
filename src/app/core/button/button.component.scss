@use "../../../styles/variables/colors";
@use "../../../styles/variables/sizes";
@use "../../../styles/variables/typography";

.button {
  background-color: colors.$primary;
  color: colors.$neutralWhite;
  padding: sizes.$buttonPadding;
  border-radius: sizes.$borderRadius;
  font-weight: typography.$semiBold;
  border: 2px solid colors.$primary;

  &:hover {
    background-color: colors.$primaryLightest;
    color: colors.$primary;
  }

  &--inverted {
    background-color: colors.$primaryLightest;
    color: colors.$primary;
  }

  &--big {
    width: 100%;
  }

  &:disabled {
    background-color: colors.$primaryLightest;
    color: colors.$primaryLight;
    &:hover {
      cursor: not-allowed;
    }
  }
}
