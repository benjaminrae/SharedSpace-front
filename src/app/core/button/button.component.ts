import {
  EventEmitter,
  Output,
  Component,
  Input,
  OnInit,
  ViewChild,
  ElementRef,
} from "@angular/core";
@Component({
  selector: "app-button",
  templateUrl: "./button.component.html",
  styleUrls: ["./button.component.scss"],
})
export class ButtonComponent implements OnInit {
  @Input() inverted!: boolean;
  @Input() big!: boolean;
  @Input() text = "Click here";
  @Input() type = "submit";
  @Input() disabled!: boolean;
  @Input() routerLink!: string;
  @Output() onClick = new EventEmitter();
  @ViewChild("button", { read: ElementRef })
  button!: ElementRef<HTMLButtonElement>;

  className!: string;

  onClickButton(event: Event) {
    event.stopPropagation();
    this.onClick.emit(event);
  }

  ngOnInit(): void {
    this.className = this.getClassName();
  }

  private getClassName() {
    return `button${this.inverted ? " button--inverted" : ""}${
      this.big ? " button--big" : ""
    }`;
  }
}
