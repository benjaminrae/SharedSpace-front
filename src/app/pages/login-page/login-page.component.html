<div class="login-page">
  <img src="../../../assets/svgs/workspace.svg" />

  <app-login-form class="login-page__form"></app-login-form>
</div>
