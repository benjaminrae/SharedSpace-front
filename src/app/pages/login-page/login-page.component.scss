.login-page {
  display: flex;
  justify-content: space-between;
  align-items: unset;
  padding-top: 2rem;
  gap: 2rem;

  & img {
    display: none;
  }

  &__form {
    flex: 1;
  }
}

@media only screen and (min-width: 800px) {
  .login-page {
    & img {
      display: block;
      flex: 1;
      max-width: 60%;
    }

    &__form {
      flex: 1;
    }
  }
}
