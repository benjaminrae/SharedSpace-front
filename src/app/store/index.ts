import { type ActionReducerMap, type MetaReducer } from "@ngrx/store";
import { environment } from "../../environments/environment";
import { locationsFeature } from "./locations-feature/locations-feature.reducer";
import { type ApplicationState } from "./types";
import { uiFeature } from "./ui-feature/ui-feature.reducer";
import { userFeature } from "./user-feature/user-feature.reducer";

export const reducers: ActionReducerMap<ApplicationState> = {
  ui: uiFeature.reducer,
  user: userFeature.reducer,
  locations: locationsFeature.reducer,
};

export const metaReducers: Array<MetaReducer<ApplicationState>> =
  !environment.production ? [] : [];
