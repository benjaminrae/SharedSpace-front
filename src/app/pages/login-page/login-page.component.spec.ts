import { HttpClient, HttpHandler } from "@angular/common/http";
import { ComponentFixture, TestBed } from "@angular/core/testing";
import { FormBuilder } from "@angular/forms";
import { provideMockStore } from "@ngrx/store/testing";
import { render, screen } from "@testing-library/angular";
import { ButtonComponent } from "../../core/button/button.component";
import { LoginFormComponent } from "../../components/login-form/login-form.component";

import { LoginPageComponent } from "./login-page.component";

const providers = [FormBuilder, HttpClient, HttpHandler, provideMockStore({})];
const declarations = [LoginFormComponent, ButtonComponent];

describe("LoginPageComponent", () => {
  let component: LoginPageComponent;
  let fixture: ComponentFixture<LoginPageComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [LoginPageComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(LoginPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it("should create", () => {
    expect(component).toBeTruthy();
  });
});

describe("Given the LoginPage component", () => {
  describe("When it is rendered", () => {
    test("Then it should show an image on the screen with name 'A man working at a desk' and a heading level 2 with 'Log in to SharedSpace'", async () => {
      const imageAltText = "A man working at a desk";
      const loginHeading = {
        level: 2,
        name: "Log in to SharedSpace",
      };

      await render(LoginPageComponent, { declarations, providers });

      const renderedHeading = screen.queryByRole("heading", loginHeading);
      const renderedImage = screen.queryByRole("img", {
        name: imageAltText,
      });

      expect(renderedHeading).toBeInTheDocument();
      expect(renderedImage).toBeInTheDocument();
    });
  });
});
