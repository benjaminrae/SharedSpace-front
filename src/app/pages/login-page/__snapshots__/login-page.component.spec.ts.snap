// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the LoginPageComponent When it is rendered and then it should match the snapshot 1`] = `
LoginPageComponent {
  "__ngContext__": 2,
}
`;
